### PR TITLE
xds: xds load report store implementation

### PR DESCRIPTION
**Pull request by voidzcy**

### CODE REVIEW

File: xds/src/main/java/io/grpc/xds/XdsClientLoadRecorder.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+    /**
+     * Generate a query count snapshot and reset counts for next snapshot.
+     */
+    ClientLoadSnapshot snapshot() {
+      ClientLoadSnapshot snapshot = new ClientLoadSnapshot();
+      long numStarted = callsStarted.get();
+      long numFinished = callsFinished.getAndSet(0);
+      // Number of calls started rolls over to next snapshot to account for calls in progress.
+      callsStarted.set(numStarted - numFinished);
```
Comment (zhangkun83) [inline xds/src/main/java/io/grpc/xds/XdsClientLoadRecorder.java]: This doesn't seem right. In the subsequent `numStarted` will have more than actual started number.

`callsInProgress` is different from the other two cumulative counters. It can use its own counter that increments at the beginning of RPC and decrements at the end. It will not be reset when generating reports. `numStarted` should just have `getAndSet(0)` like `numFinished`.

File: xds/src/main/java/io/grpc/xds/XdsLoadReportStore.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+import io.grpc.LoadBalancer.PickResult;
+import io.grpc.Metadata;
+import io.grpc.xds.XdsClientLoadRecorder.ClientLoadCounter;
+import java.util.HashMap;
+import java.util.Map;
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * An {@link XdsLoadReportStore} instance holds the client side load stats for a cluster. Methods
+ * should be called in a synchronized context.
```
Comment (zhangkun83) [inline xds/src/main/java/io/grpc/xds/XdsLoadReportStore.java]: This class has to be thread-safe, because `interceptPickResult()` is per-RPC, which come in concurrently.

File: core/src/main/java/io/grpc/util/ForwardingClientStreamTracer.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+ */
+
+package io.grpc.util;
+
+import com.google.common.base.MoreObjects;
+import io.grpc.ClientStreamTracer;
+import io.grpc.Metadata;
+import io.grpc.Status;
+
+public abstract class ForwardingClientStreamTracer extends ClientStreamTracer {
```
Comment (zhangkun83) [inline core/src/main/java/io/grpc/util/ForwardingClientStreamTracer.java]: This file and its test should be in a separate PR.

File: xds/src/main/java/io/grpc/xds/XdsClientLoadRecorder.java (hunk 1 above)
Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/XdsClientLoadRecorder.java]: Yes, I agree. This is slightly different from what the design doc suggests for implementation, but it is a correct way to count the calls. So we would actually do not need to count calls started. Instead, we just need to have counters for `callsInProgress` (+ at start and - at complete), `callsFinished` (+ at complete) and `callsFailed` (+ when status is not ok at complete). These three are enough for computing `total_successful_request`, `total_request_in_progress` and `total_error_requests`.

File: core/src/main/java/io/grpc/util/ForwardingClientStreamTracer.java (hunk 3 above)
Comment (voidzcy) [reply core/src/main/java/io/grpc/util/ForwardingClientStreamTracer.java]: ACK.

Comment (voidzcy) [reply core/src/main/java/io/grpc/util/ForwardingClientStreamTracer.java]: #5589

File: xds/src/main/java/io/grpc/xds/XdsClientLoadRecorder.java [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+import io.grpc.ClientStreamTracer;
+import io.grpc.ClientStreamTracer.StreamInfo;
+import io.grpc.Metadata;
+import io.grpc.Status;
+import io.grpc.util.ForwardingClientStreamTracer;
+import java.util.concurrent.atomic.AtomicLong;
+import javax.annotation.concurrent.ThreadSafe;
+
+@ThreadSafe
+public class XdsClientLoadRecorder extends ClientStreamTracer.Factory {
```
Comment (zhangkun83) [inline xds/src/main/java/io/grpc/xds/XdsClientLoadRecorder.java]: All classes should be final unless they are intended to be overridden.

File: xds/src/main/java/io/grpc/xds/XdsClientLoadRecorder.java [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    counter.callsStarted.getAndIncrement();
+    final ClientStreamTracer delegateTracer = delegate.newClientStreamTracer(info, headers);
+    return new StreamTracer(delegateTracer);
+  }
+
+  /**
+   * A {@link ClientLoadSnapshot} represents a snapshot of {@link ClientLoadCounter} to be sent as
+   * part of {@link io.envoyproxy.envoy.api.v2.endpoint.ClusterStats} to the balancer.
+   */
+  static class ClientLoadSnapshot {
```
Comment (zhangkun83) [inline xds/src/main/java/io/grpc/xds/XdsClientLoadRecorder.java]: This class needs to be final, and all fields final.

File: xds/src/main/java/io/grpc/xds/XdsLoadReportStore.java [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+    this.localityLoadCounters = checkNotNull(localityLoadCounters, "localityLoadCounters");
+  }
+
+  /**
+   * Generates a {@link ClusterStats} containing load stats in locality granularity.
+   */
+  ClusterStats generateLoadReport(Duration interval) {
+    ClusterStats.Builder statsBuilder = ClusterStats.newBuilder().setClusterName(clusterName)
+        .setLoadReportInterval(interval);
+    for (Locality locality : localityLoadCounters.keySet()) {
```
Comment (zhangkun83) [inline xds/src/main/java/io/grpc/xds/XdsLoadReportStore.java]: Iterate on `localityLoadCounters.entrySet()` instead. Because `localityLoadCounters` should be a concurrent map, the key could be removed before you call `get()`.

File: xds/src/main/java/io/grpc/xds/XdsLoadReportStore.java [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+
+/**
+ * An {@link XdsLoadReportStore} instance holds the client side load stats for a cluster. Methods
+ * should be called in a synchronized context.
+ */
+@NotThreadSafe
+class XdsLoadReportStore {
+
+  private final String clusterName;
+  private final Map<Locality, XdsClientLoadRecorder.ClientLoadCounter> localityLoadCounters;
```
Comment (zhangkun83) [inline xds/src/main/java/io/grpc/xds/XdsLoadReportStore.java]: There should be a way to remove a locality from the map if it no longer has traffic, probably because xDS traffic director sent a new list of Localities that don't include it any more.

This may be done when generating a report. If all counters in the report are 0, the Locality may be removed. Doing this with concurrency in mind would require some tricks.

Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/XdsLoadReportStore.java]: I added a `discardClientLoadCounter` method, which can be called in locality update for removed localities. For adding the counter, I kept it as it is, done when trying to intercept the `PickResult` while the counter is missing.

Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/XdsLoadReportStore.java]: Nvm. I will add a `createClientLoadCounter` method to be called in locality update for newly assigned localities. Current approach might have a race between counter being added/discarded, although it does no harm.

File: xds/src/main/java/io/grpc/xds/XdsLoadReportStore.java [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+        public ClientStreamTracer newClientStreamTracer(StreamInfo info, Metadata headers) {
+          return new ClientStreamTracer() {
+          };
+        }
+      };
+    }
+
+    // TODO (chengyuanzhang): use a createClientLoadCounter method instead better?
+    XdsClientLoadRecorder.ClientLoadCounter counter = localityLoadCounters
+        .putIfAbsent(locality, new ClientLoadCounter());
```
Comment (zhangkun83) [inline xds/src/main/java/io/grpc/xds/XdsLoadReportStore.java]: There should be an `addClientLoadCounter()` to go with `discardClientLoadCounter()`. Or better, just `addLocality()` and `removeLocality()` to give out less implementation detail. The callers of `XdsLoadReportStore` doesn't have to know about any "counter".

You don't have to have an explicit `addLocality()` per se, but I would recommend you do so, because:
1. It's easier, because then the mutation of the map can be done completely in the sync context, and there is no need for `putIfAbsent()`. 
2. It's less error-prone. If you add to the map in `interceptPickResult()`, which is on the data path (per-RPC, application thread), it may race with `removeLocality()` which is on the control path (sync context). You would need to be careful about the ordering of calling `removeLocality()` vs. updating the picker, to make sure a removed locality is never passed to `interceptPickResult()` after it's removed from the map. On the other hand, if you have `addLocality()` to add to the map, even if a removed locality is passed to `interceptPickResult()`, you can simply ignore it.

Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/XdsLoadReportStore.java]: ACK.

File: xds/src/main/java/io/grpc/xds/XdsLoadReportStore.java [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+    return statsBuilder.build();
+  }
+
+  /**
+   * Create a {@link ClientLoadCounter} for the provided locality if not present in this {@link
+   * XdsLoadReportStore}. This method needs to be called at locality updates for newly assigned
+   * localities in balancer discovery responses.
+   */
+  void addLocality(Locality locality) {
+    localityLoadCounters.putIfAbsent(locality, new ClientLoadCounter());
```
Comment (zhangkun83) [inline xds/src/main/java/io/grpc/xds/XdsLoadReportStore.java]: The xDS balancer should be able to correctly track which localities are new. It should be considered a bug if the xDS policy call this with a locality that's already here. We should call `put()` and assert that the return value is `null` to expose such bug.

File: xds/src/main/java/io/grpc/xds/XdsLoadReportStore.java [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+   */
+  void removeLocality(Locality locality) {
+    localityLoadCounters.remove(locality);
+  }
+
+  /**
+   * Intercepts a in-locality PickResult with load recording {@link ClientStreamTracer.Factory}.
+   */
+  PickResult interceptPickResult(PickResult pickResult, Locality locality) {
+    if (!pickResult.getStatus().isOk() || !localityLoadCounters.containsKey(locality)) {
```
Comment (zhangkun83) [inline xds/src/main/java/io/grpc/xds/XdsLoadReportStore.java]: nit: I would split this into two if-branches, with the second being:
```java
XdsClientLoadRecorder.ClientLoadCounter counter = localityLoadCounters.get(locality);
if (counter == null) {
  return;
}
```

Then you don't call `get()` again below. This has a minor performance advantage that may be negligible. It also helps with readability, because it's very obvious `counter` is not null from this point, but it's less obvious `localityLoadCounters.get(locality)` is not null.

File: xds/src/main/java/io/grpc/xds/XdsLoadReportStore.java [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+  /**
+   * Intercepts a in-locality PickResult with load recording {@link ClientStreamTracer.Factory}.
+   */
+  PickResult interceptPickResult(PickResult pickResult, Locality locality) {
+    if (!pickResult.getStatus().isOk()) {
+      return pickResult;
+    }
+    ClientStreamTracer.Factory originFactory = pickResult.getStreamTracerFactory();
+    if (originFactory == null) {
+      originFactory = new ClientStreamTracer.Factory() {
```
[the remark this answers is not in the document]
Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/XdsLoadReportStore.java]: ACK.

File: xds/src/main/java/io/grpc/xds/XdsLoadReportStore.java [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+      };
+    }
+
+    XdsClientLoadRecorder.ClientLoadCounter counter = localityLoadCounters.get(locality);
+    XdsClientLoadRecorder recorder = new XdsClientLoadRecorder(counter, originFactory);
+    return PickResult.withSubchannel(pickResult.getSubchannel(), recorder);
+  }
+
+  /**
+   * Record that a request has been dropped by drop overload policy with the provided category
```
Comment (zhangkun83) [inline xds/src/main/java/io/grpc/xds/XdsLoadReportStore.java]: Please find the answer with the author of the document and have them write it down before merging this PR.

File: xds/src/main/java/io/grpc/xds/XdsLoadReportStore.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (zhangkun83) [reply xds/src/main/java/io/grpc/xds/XdsLoadReportStore.java]: The factory could be static too.

File: xds/src/main/java/io/grpc/xds/XdsLoadReportStore.java [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+    localityLoadCounters.putIfAbsent(locality, new ClientLoadCounter());
+  }
+
+  /**
+   * Discard the {@link ClientLoadCounter} for the provided locality if owned by this {@link
+   * XdsLoadReportStore} to avoid map size growing infinitely. To be called at locality updates when
+   * the provided locality is no longer considered in balancer discovery response.s
+   */
+  void removeLocality(Locality locality) {
+    localityLoadCounters.remove(locality);
```
Comment (zhangkun83) [inline xds/src/main/java/io/grpc/xds/XdsLoadReportStore.java]: You test code reminds me of long-running RPCs. If a locality is removed by xDS balancer while there are still running RPCs for it and they don't finish soon, there are still substantial load data for this locality to report. You may want to confirm with the design author about this, but my guess is that you don't want to remove it right away. If that's the case, a possible solution is to add a boolean "active" to `ClientLoadCounter`, so that `removeLocality()` only sets the boolean to false, `addLocality()` makes sure the boolean is true, and `generateLoadReport()` removes the entry if all counters are zero and the boolean is false.

Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/XdsLoadReportStore.java]: When calling `interceptPickResult` with an existing but inactive `ClientLoadCounter`, which means we want to start tracking a backend in the locality that is no longer exposed by the balancer, should we proceed?

File: xds/src/main/java/io/grpc/xds/XdsLoadReportStore.java (hunk 10 above)
Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/XdsLoadReportStore.java]: ACK.

File: xds/src/main/java/io/grpc/xds/XdsLoadReportStore.java (hunk 13 above)
Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/XdsLoadReportStore.java]: ACK.

File: xds/src/test/java/io/grpc/xds/XdsLoadReportStoreTest.java [hunk 14]
```diff
@@ -0,0 +1,261 @@
...
+
+  private ClusterStats buildClusterStats(Duration interval,
+      List<UpstreamLocalityStats> upstreamLocalityStatsList) {
+    return ClusterStats.newBuilder().setClusterName(SERVICE_NAME)
+        .addAllUpstreamLocalityStats(upstreamLocalityStatsList)
+        .setLoadReportInterval(interval)
+        .build();
+  }
+
+  private void assertClusterStatsEqual(ClusterStats stats1, ClusterStats stats2) {
```
Comment (zhangkun83) [inline xds/src/test/java/io/grpc/xds/XdsLoadReportStoreTest.java]: Why can't you use the equality of `ClusterStats`?

File: xds/src/test/java/io/grpc/xds/XdsLoadReportStoreTest.java [hunk 15]
```diff
@@ -0,0 +1,261 @@
...
+            .build();
+    loadStore.addLocality(locality2);
+    PickResult pickResult3 = PickResult.withSubchannel(fakeSubchannel);
+    PickResult interceptedPickResult3 = loadStore.interceptPickResult(pickResult3, locality2);
+    assertThat(localityLoadCounters).containsKey(locality2);
+    assertThat(localityLoadCounters).hasSize(2);
+    interceptedPickResult3
+        .getStreamTracerFactory()
+        .newClientStreamTracer(STREAM_INFO, new Metadata());
+    List<UpstreamLocalityStats> upstreamLocalityStatsList = new ArrayList<>();
```
Comment (zhangkun83) [inline xds/src/test/java/io/grpc/xds/XdsLoadReportStoreTest.java]: You can use `Arrays.asList()` to save a few characters.

File: xds/src/test/java/io/grpc/xds/XdsLoadReportStoreTest.java (hunk 14 above)
Comment (voidzcy) [reply xds/src/test/java/io/grpc/xds/XdsLoadReportStoreTest.java]: `ClusterStats` contains a list of `UpstreamLocalityStats`, for which the ordering we do not know. Two lists of `UpstreamLocalityStats` containing same instances should be considered "equal".

File: xds/src/main/java/io/grpc/xds/XdsLoadReportStore.java [hunk 16]
```diff
@@ -0,0 +1,166 @@
...
+    XdsClientLoadRecorder recorder = new XdsClientLoadRecorder(counter, originFactory);
+    return PickResult.withSubchannel(pickResult.getSubchannel(), recorder);
+  }
+
+  /**
+   * Record that a request has been dropped by drop overload policy with the provided category
+   * instructed by the remote balancer.
+   */
+  void recordDroppedRequest(String category) {
+    AtomicLong counter = dropCounters.putIfAbsent(category, new AtomicLong());
```
[the remark this answers is not in the document]
Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/XdsLoadReportStore.java]: For counting dropped requests in xDS, I am thinking that we could have a `DroppedPickResult` that extends `PickResult` with an extra field `reason` for the category of this drop belongs to. `PickResult` already has a method `isDrop` that tells if the result is a drop. Then we could gracefully handle the dropped requests counting in `interceptPickResult` method by first checking `pickResult.isDrop()`, if it is a drop, we cast it to `DroppedPickResult`, call `getReason()` on it to get the drop reason and update the corresponding counter in `dropCounters` map. With this approach, the xDS policy does not need to call `recordDroppedRequest` to count it. The only ugliness of this approach is the down casting, but in this case, I think it is totally fine.

File: xds/src/test/java/io/grpc/xds/XdsLoadReportStoreTest.java [hunk 17]
```diff
@@ -0,0 +1,261 @@
...
+  @Before
+  public void setUp() {
+    MockitoAnnotations.initMocks(this);
+    localityLoadCounters = new ConcurrentHashMap<>();
+    dropCounters = new ConcurrentHashMap<>();
+    loadStore = new XdsLoadReportStore(SERVICE_NAME, localityLoadCounters, dropCounters);
+  }
+
+  @Test
+  public void testUntrackedLocalityNoLoadRecording() {
```
[the remark this answers is not in the document]
Comment (voidzcy) [reply xds/src/test/java/io/grpc/xds/XdsLoadReportStoreTest.java]: What's the convention for us? I saw in our codebase, some tests are prepended with `test` while others are not. It'd be better to stick on one side. (I am leaning towards always prepend test methods with `test`:)).

File: xds/src/test/java/io/grpc/xds/XdsLoadReportStoreTest.java (hunk 15 above)
Comment (voidzcy) [reply xds/src/test/java/io/grpc/xds/XdsLoadReportStoreTest.java]: ACK.

File: xds/src/test/java/io/grpc/xds/XdsLoadReportStoreTest.java [hunk 18]
```diff
@@ -0,0 +1,261 @@
...
+    when(mockFactory
+        .newClientStreamTracer(any(ClientStreamTracer.StreamInfo.class), any(Metadata.class)))
+        .thenReturn(mockTracer);
+    PickResult pickResult = PickResult.withSubchannel(fakeSubchannel, mockFactory);
+    PickResult interceptedPickResult = loadStore.interceptPickResult(pickResult, TEST_LOCALITY);
+    Metadata metadata = new Metadata();
+    interceptedPickResult.getStreamTracerFactory().newClientStreamTracer(STREAM_INFO, metadata)
+        .streamClosed(Status.OK);
+    ArgumentCaptor<ClientStreamTracer.StreamInfo> streamInfoArgumentCaptor = ArgumentCaptor
+        .forClass(ClientStreamTracer.StreamInfo.class);
```
[the remark this answers is not in the document]
Comment (voidzcy) [reply xds/src/test/java/io/grpc/xds/XdsLoadReportStoreTest.java]: ACK.

File: xds/src/main/java/io/grpc/xds/XdsLoadReportStore.java [hunk 19]
```diff
@@ -0,0 +1,166 @@
...
+              .setTotalSuccessfulRequests(snapshot.callsSucceed)
+              .setTotalErrorRequests(snapshot.callsFailed)
+              .setTotalRequestsInProgress(snapshot.callsInProgress));
+      // Discard counters for localities that are no longer exposed by the remote balancer and
+      // no RPCs ongoing.
+      if (!entry.getValue().isActive() && snapshot.callsInProgress == 0) {
+        localityLoadCounters.remove(entry.getKey());
+      }
+    }
+    for (Map.Entry<String, AtomicLong> entry : dropCounters.entrySet()) {
```
[the remark this answers is not in the document]
Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/XdsLoadReportStore.java]: ACK.

File: xds/src/test/java/io/grpc/xds/XdsLoadReportStoreTest.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (zhangkun83) [reply xds/src/test/java/io/grpc/xds/XdsLoadReportStoreTest.java]: The convention is to name the method after the tested feature or behavior. For example, this method would be named "loadNotRecordedForUntrackedLocality()".

File: xds/src/main/java/io/grpc/xds/XdsLoadReportStore.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (zhangkun83) [reply xds/src/main/java/io/grpc/xds/XdsLoadReportStore.java]: `PickResult` is final so it's not ready to be extended. `isDrop()` is on it because that signal is used by gRPC library to determine its retry eligibility. It's not the case for the drop category. Having drop-counting in `XdsLoadReportStore` instead of the xDS policy doesn't sound a good enough reason to make the API change on `PickResult`.

File: xds/src/main/java/io/grpc/xds/XdsLoadReportStore.java (hunk 9 above)
Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/XdsLoadReportStore.java]: ACK.
